source/adapter.screenshare.js

```diff
@@ -26,147 +55,284 @@ AdapterJS.defineMediaSourcePolyfill = function () {
     baseGetUserMedia = window.navigator.getUserMedia;
 
     navigator.getUserMedia = function (constraints, successCb, failureCb) {
+      // Append checks for overrides as these are mandatory
+      // Browsers (not Firefox since they went Promise based) does these checks and they can be quite useful
+      if (!(constraints && typeof constraints === 'object')) {
+        throw new Error('GetUserMedia: (constraints, .., ..) argument required');
+      } else if (typeof successCb !== 'function') {
+        throw new Error('GetUserMedia: (.., successCb, ..) argument required');
+      } else if (typeof failureCb !== 'function') {
+        throw new Error('GetUserMedia: (.., .., failureCb) argument required');
+      }
 
-      if (constraints && constraints.video && !!constraints.video.mediaSource) {
-        // intercepting screensharing requests
+      // Prevent accessing property from Boolean errors
+      if (constraints.video && typeof constraints.video === 'object' &&
+        constraints.video.hasOwnProperty('mediaSource')) {
+        var updatedConstraints = clone(constraints);
+        // See: http://fluffy.github.io/w3c-screen-share/#screen-based-video-constraints
+        // See also: https://bugzilla.mozilla.org/show_bug.cgi?id=1037405
+        var mediaSourcesList = ['screen', 'window', 'application', 'browser', 'camera'];
+        var useExtensionErrors = ['NotAllowedError', 'PermissionDeniedError', 'SecurityError'];
+
+        // Obtain first item in array if array is provided
+        if (Array.isArray(updatedConstraints.video.mediaSource)) {
+          var i = 0;
+          while (i < updatedConstraints.video.mediaSource.length) {
+            if (mediaSourcesList.indexOf(updatedConstraints.video.mediaSource[i]) > -1) {
+              updatedConstraints.video.mediaSource = updatedConstraints.video.mediaSource[i];
+              break;
+            }
+            i++;
+          }
+          updatedConstraints.video.mediaSource = typeof updatedConstraints.video.mediaSource === 'string' ?
+            updatedConstraints.video.mediaSource : null;
+        }
 
         // Invalid mediaSource for firefox, only "screen" and "window" are supported
-        if (constraints.video.mediaSource !== 'screen' && constraints.video.mediaSource !== 'window') {
+        if (mediaSourcesList.indexOf(updatedConstraints.video.mediaSource) === -1) {
```

Comment:
@miniruwan it is checked here for `['screen', 'window', 'application', 'browser', 'camera']`